src/nat-manager.js

```diff
@@ -0,0 +1,137 @@
+'use strict'
+
+const NatAPI = require('@motrix/nat-api')
+const debug = require('debug')
+const promisify = require('promisify-es6')
+const Multiaddr = require('multiaddr')
+const log = debug('libp2p:nat')
+log.error = debug('libp2p:nat:error')
+const { isBrowser } = require('ipfs-utils/src/env')
+const retry = require('p-retry')
+const isPrivateIp = require('private-ip')
+const pkg = require('../package.json')
+
+class NatManager {
+  /**
+   * @class
+   * @param {object} options
+   * @param {PeerId} options.peerId - The peer ID of the current node
+   * @param {TransportManager} options.transportManager - A transport manager
+   * @param {AddressManager} options.addressManager - An address manager
+   */
+  constructor ({ peerId, addressManager, transportManager, ...options }) {
+    this._peerId = peerId
+    this._addressManager = addressManager
+    this._transportManager = transportManager
+
+    this._enabled = options.enabled
+    this._externalIp = options.externalIp
+    this._options = {
+      description: options.description || `${pkg.name}@${pkg.version} ${this._peerId}`,
+      ttl: options.ttl || 7200,
+      autoUpdate: options.autoUpdate || true,
+      gateway: options.gateway,
+      enablePMP: Boolean(options.pmp && options.pmp.enabled)
+    }
+  }
+
+  /**
+   * Starts the NAT manager
+   */
+  start () {
+    if (isBrowser || !this._enabled) {
+      return
+    }
+
+    // done async to not slow down startup
+    this._start().catch((err) => {
+      // hole punching errors are non-fatal
+      log.error(err)
+    })
+  }
+
+  async _start () {
+    const addrs = this._transportManager.getAddrs()
+
+    for (const addr of addrs) {
+      // try to open uPnP ports for each thin waist address
+      const { family, host, port, transport } = addr.toOptions()
+
+      if (!addr.isThinWaistAddress() || transport !== 'tcp') {
+        // only bare tcp addresses
+        continue
+      }
+
+      if (host === '127.0.0.1' || host === '::1') {
+        // ignore loopback addresses
+        continue
+      }
+
+      if (family !== 'ipv4') {
+        // ignore ipv6
+        continue
+      }
+
+      const client = this._getClient()
+      const publicIp = this._externalIp || await client.externalIp()
+
+      if (isPrivateIp(publicIp)) {
+        throw new Error(`${publicIp} is private - please set config.nat.externalIp to an externally routable IP or ensure you are not behind a double NAT`)
+      }
+
+      log(`opening uPnP connection from ${publicIp}:${port} to ${host}:${port}`)
+
+      await client.map({
+        publicPort: port,
+        privatePort: port,
+        protocol: transport.toUpperCase()
+      })
+
+      this._addressManager.announce.add(Multiaddr.fromNodeAddress({
+        family: 'IPv4',
+        address: publicIp,
+        port
+      }, transport))
```

Comment:
The mapping code will only ever run once per TCP address we have, correct? The NAT Manager has an autoUpdate option, so I'm assuming the client itself will automatically update addresses if they change. Is that the case? If so, we're not updating our nat addresses at all.

If this does need to update over time ( I think it does ), how we're setting the addresses will need to change, as an add isn't going to be sufficient. Also, this add won't trigger an Identify Push. The consideration, https://github.com/libp2p/js-libp2p/tree/master/src/address-manager#modify-announce-addresses, is something we need to account for with this. cc @vasco-santos. At the bare minimum we'll likely need to leverage the work done for 0.30 and update our address record to cause the trigger of events in the subsystems, https://github.com/libp2p/js-libp2p/blob/0.30.x/src/transport-manager.js#L164.

Reply:
> The NAT Manager has an autoUpdate option

Maybe this is poorly named and should be `autoRefresh` or `keepAlive` or something - it means to periodically tell the router to keep the UPnP mapping alive, otherwise it'll be removed once the TTL is reached.

Reply:
`keepAlive` sounds like the way to go.

For adding the address I think we should add this as an observed address, and add that base logic to the address manager. We'll eventually want to factor in confidence to these, but for now we can just add Nat Manager addresses to it. The announce address logic is being updated to reflect go-libp2p, so if any announce address is provided it will cause the other listen addresses to be hidden which is not what we want to happen. @vasco-santos  what are your thoughts on this.

Reply:
I agree on getting the observedAddresses logic into the AddressManager would be the best way to go. We should gather also the observed addresses from [identify](https://github.com/libp2p/js-libp2p/blob/v0.29.3/src/identify/index.js#L207).

For creating the announce addresses, I think we should keep things as `0.30` currently does. This is, if we have user provided announce addresses only these addresses should be announced. If we do not have announce addresses, we will return the listenAddr + observedAddr

Regarding updating the self peer record and doing identify push to update the connected peers about the self multiaddrs changes, I think that we can make `AddressManager` an event emitter to trigger `updateSelfPeerRecord`. This would also unblock the programatic API for the AddressManager.

Reply:
As I see it there are a number of asks here:

- [x] Store NAT punched addresses as observed addresses using the address manager
- [ ] Store observed addresses from identify using the address manager
- [x] Make AddressManager an event emitter that triggers updateSelfPeerRecord
- [ ] Send configured observed addresses plus either configured announceAddrs or the listenAddr during identify
- [ ] Invoke consumePeerRecord when addresses change

Reply:
@achingbrain 

>  Send configured observed addresses plus either configured announceAddrs or the listenAddr during identify
>  Invoke consumePeerRecord when addresses change

Your PR already does this. Identify listens for events of the PeerStore when an address change. Considering that you trigger the creation of a new signed peer record when new observed addresses is added, this will call consumePeerRecord which will emit the address change (if in fact it is a new address for the peer). This will trigger identify push to happens.

The "concern" here is when using announceAddresses. This is because we create a signed peer record by fetching `libp2p.multiaddrs` which are the addresses that a peer intends to let other peers know of. So, according to the current logic, if we have `announceAddresses` configured, `libp2p.multiaddrs` will always be the announceAddresses. Which makes, observed addresses never used. When a new signed peer record is created, it will include exactly the same addresses as before (the announce ones).

Reply:
> Your PR already does this

Yeah, I got distracted while double checking the PR for the checkboxes..